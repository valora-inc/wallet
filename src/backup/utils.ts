import {
  CELO_DERIVATION_PATH_BASE,
  generateKeys,
  MnemonicLanguages,
} from '@celo/cryptographic-utils'
import CryptoJS from 'crypto-js'
import { useAsync } from 'react-async-hook'
import * as bip39 from 'react-native-bip39'
import { showError } from 'src/alert/actions'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { getPassword } from 'src/pincode/authentication'
import { useDispatch, useSelector } from 'src/redux/hooks'
import { removeStoredItem, retrieveStoredItem, storeItem } from 'src/storage/keychain'
import Logger from 'src/utils/Logger'
import { ETHEREUM_DERIVATION_PATH } from 'src/web3/consts'
import { currentAccountSelector } from 'src/web3/selectors'

const TAG = 'Backup/utils'

const MNEMONIC_STORAGE_KEY = 'mnemonic'

export async function generateKeysFromMnemonic(mnemonic: string) {
  const wordCount = countMnemonicWords(mnemonic)
  const derivationPath = wordCount === 24 ? CELO_DERIVATION_PATH_BASE : ETHEREUM_DERIVATION_PATH
  return generateKeys(mnemonic, undefined, undefined, undefined, bip39, derivationPath)
}

export function getMnemonicLanguage(language: string | null) {
  switch (language?.slice(0, 2)) {
    case 'es': {
      return MnemonicLanguages.spanish
    }
    case 'pt': {
      return MnemonicLanguages.portuguese
    }
    default: {
      return MnemonicLanguages.english
    }
  }
}

export async function storeMnemonic(mnemonic: string, account: string | null, password?: string) {
  if (!account) {
    throw new Error('Account not yet initialized')
  }
  const passwordToUse = password ?? (await getPassword(account))
  const encryptedMnemonic = await encryptMnemonic(mnemonic, passwordToUse)
  return storeItem({ key: MNEMONIC_STORAGE_KEY, value: encryptedMnemonic })
}

export async function clearStoredMnemonic() {
  await removeStoredItem(MNEMONIC_STORAGE_KEY)
}

export async function getStoredMnemonic(
  account: string | null,
  password?: string
): Promise<string | null> {
  try {
    if (!account) {
      throw new Error('Account not yet initialized')
    }

    Logger.debug(TAG, 'Checking keystore for mnemonic')
    const encryptedMnemonic = await retrieveStoredItem(MNEMONIC_STORAGE_KEY)
    if (!encryptedMnemonic) {
      throw new Error('No mnemonic found in storage')
    }

    const passwordToUse = password ?? (await getPassword(account))
    return decryptMnemonic(encryptedMnemonic, passwordToUse)
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}

export function onGetMnemonicFail(viewError: (error: ErrorMessages) => void, context?: string) {
  viewError(ErrorMessages.FAILED_FETCH_MNEMONIC)
  ValoraAnalytics.track(OnboardingEvents.backup_error, {
    error: 'Failed to retrieve Recovery Phrase',
    context,
  })
}

export function useAccountKey(): string | null {
  const dispatch = useDispatch()
  const account = useSelector(currentAccountSelector)
  const asyncAccountKey = useAsync(getStoredMnemonic, [account])

  if (!asyncAccountKey || asyncAccountKey.error) {
    onGetMnemonicFail((error) => dispatch(showError(error)), 'useAccountKey')
  }

  return asyncAccountKey.result || null
}

export function countMnemonicWords(phrase: string): number {
  return [...phrase.trim().split(/\s+/)].length
}

// Because of a RN bug, we can't fully clean the text as the user types
// https://github.com/facebook/react-native/issues/11068
export function formatBackupPhraseOnEdit(phrase: string) {
  return phrase.replace(/\s+/gm, ' ')
}

export function isValidBackupPhrase(phrase: string) {
  const allowedPhraseLengths = [12, 24]
  const phraseLength = countMnemonicWords(formatBackupPhraseOnEdit(phrase))
  return !!phrase && allowedPhraseLengths.includes(phraseLength)
}

export async function encryptMnemonic(phrase: string, password: string) {
  return CryptoJS.AES.encrypt(phrase, password).toString()
}

export async function decryptMnemonic(encryptedMnemonic: string, password: string) {
  const bytes = CryptoJS.AES.decrypt(encryptedMnemonic, password)
  return bytes.toString(CryptoJS.enc.Utf8)
}
