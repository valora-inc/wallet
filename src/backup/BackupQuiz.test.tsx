import { fireEvent, render, waitFor } from '@testing-library/react-native'
import * as React from 'react'
import { Provider } from 'react-redux'
import BackupQuiz, { BackupQuiz as BackupQuizRaw, navOptionsForQuiz } from 'src/backup/BackupQuiz'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import MockedNavigator from 'test/MockedNavigator'
import { createMockStore, getMockI18nProps, getMockStackScreenProps } from 'test/utils'
import { mockAccount, mockMnemonic } from 'test/values'

jest.mock('lodash', () => ({
  ...(jest.requireActual('lodash') as any),
  shuffle: jest.fn((array) => array),
}))

const mockGetStoredMnemonic = jest.fn(() => mockMnemonic)
jest.mock('src/backup/utils', () => ({
  ...(jest.requireActual('src/backup/utils') as any),
  getStoredMnemonic: () => mockGetStoredMnemonic(),
}))

const mockScreenProps = getMockStackScreenProps(Screens.BackupQuiz)

describe('BackupQuiz', () => {
  const store = createMockStore()

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', async () => {
    const { getByTestId, toJSON } = render(
      <Provider store={store}>
        <BackupQuiz {...mockScreenProps} />
      </Provider>
    )
    await waitFor(() => getByTestId('selected-word-0'))
    expect(toJSON()).toMatchSnapshot()
  })

  it('Cancel navigates correctly when not in account removal', () => {
    const { getByText, getByTestId } = render(
      <Provider store={store}>
        <MockedNavigator component={BackupQuiz} options={navOptionsForQuiz} />
      </Provider>
    )

    fireEvent.press(getByTestId('CancelButton'))
    expect(getByText('cancelDialog.title')).toBeTruthy()
    expect(getByText('cancelDialog.body')).toBeTruthy()
  })

  it('Cancel navigates correctly when in account removal', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <MockedNavigator
          component={BackupQuiz}
          params={{ isAccountRemoval: true }}
          options={navOptionsForQuiz}
        />
      </Provider>
    )

    fireEvent.press(getByTestId('CancelButton'))
    expect(navigate).toBeCalledWith(Screens.Settings)
  })

  describe('when word is pressed', () => {
    it('removes it from the options adds it to chosen words', async () => {
      const mockSetBackupCompleted = jest.fn()
      const { getByTestId, getByText } = render(
        <Provider store={store}>
          <BackupQuizRaw
            {...getMockStackScreenProps(Screens.BackupQuiz)}
            setBackupCompleted={mockSetBackupCompleted}
            showError={jest.fn()}
            account={mockAccount}
            {...getMockI18nProps()}
          />
        </Provider>
      )

      await waitFor(() => getByTestId('selected-word-0'))

      expect(getByTestId('selected-word-0').props.children).toEqual(1)

      const words = mockMnemonic.split(' ')
      const firstWord = words[0]
      const secondWord = words[1]

      fireEvent.press(getByText(firstWord))

      await waitFor(() =>
        getByTestId('selected-word-0').find((node) => node.props.children === firstWord)
      )

      expect(getByTestId('selected-word-0').props.children).toEqual(firstWord)

      expect(getByText(secondWord)).toBeTruthy()
    })
  })

  it.each([{ isAccountRemoval: false }, { isAccountRemoval: true }])(
    'can complete the quiz correctly (account removal: $isAccountRemoval)',
    async ({ isAccountRemoval }) => {
      jest.useFakeTimers()
      const mockSetBackupCompleted = jest.fn()
      const { getByText, getByTestId } = render(
        <Provider store={store}>
          <BackupQuizRaw
            {...getMockStackScreenProps(Screens.BackupQuiz, { isAccountRemoval })}
            setBackupCompleted={mockSetBackupCompleted}
            showError={jest.fn()}
            account={mockAccount}
            {...getMockI18nProps()}
          />
        </Provider>
      )
      for (const word of mockMnemonic.split(' ')) {
        await waitFor(() => getByText(word))
        fireEvent.press(getByText(word))
      }

      fireEvent.press(getByTestId('QuizSubmit'))
      jest.advanceTimersByTime(2000)
      expect(mockSetBackupCompleted).toHaveBeenCalled()
      expect(navigate).toHaveBeenCalledWith(Screens.BackupComplete, { isAccountRemoval })
    }
  )
})
