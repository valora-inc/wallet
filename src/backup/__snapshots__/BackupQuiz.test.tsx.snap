// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackupQuiz renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    {
      "flex": 1,
      "justifyContent": "space-between",
      "paddingBottom": 24,
      "paddingHorizontal": 24,
    }
  }
>
  <RNCSafeAreaView
    edges={
      [
        "bottom",
        "left",
        "right",
        "top",
      ]
    }
    style={
      {
        "height": 35,
        "position": "absolute",
        "right": 10,
        "top": 10,
        "width": 60,
        "zIndex": 100,
      }
    }
  >
    <Text
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
         
    </Text>
  </RNCSafeAreaView>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
        "paddingTop": 24,
      }
    }
  >
    <View>
      <View
        style={
          {
            "alignSelf": "center",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "justifyContent": "center",
            "maxWidth": 288,
          }
        }
      >
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-0"
          >
            1
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-1"
          >
            2
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-2"
          >
            3
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-3"
          >
            4
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-4"
          >
            5
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-5"
          >
            6
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-6"
          >
            7
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-7"
          >
            8
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-8"
          >
            9
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-9"
          >
            10
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-10"
          >
            11
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-11"
          >
            12
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-12"
          >
            13
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-13"
          >
            14
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-14"
          >
            15
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-15"
          >
            16
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-16"
          >
            17
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-17"
          >
            18
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-18"
          >
            19
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-19"
          >
            20
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-20"
          >
            21
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-21"
          >
            22
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-22"
          >
            23
          </Text>
        </View>
        <View
          style={
            [
              {
                "borderColor": "#E6E6E6",
                "borderRadius": 100,
                "borderWidth": 1,
                "marginHorizontal": 3,
                "marginVertical": 4,
                "minWidth": 55,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#666666",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": undefined,
                "textAlign": "center",
              }
            }
            testID="selected-word-23"
          >
            24
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
              "marginTop": 20,
              "textAlign": "center",
            }
          }
        >
          backupQuizWordCount, {"ordinal":"ordinals.1"}
          <Text
            style={
              {
                "color": "#2E3338",
                "fontFamily": "Inter-Medium",
                "fontSize": 16,
                "lineHeight": 22,
                "marginTop": 25,
                "textAlign": "center",
              }
            }
          >
            X
          </Text>
        </Text>
        <View
          style={
            {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "center",
              "marginTop": 24,
            }
          }
        >
          <View
            style={
              {
                "borderColor": "#1AB775",
                "borderRadius": 100,
                "borderWidth": 1.5,
                "marginHorizontal": 4,
                "marginVertical": 4,
                "overflow": "hidden",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              focusable={true}
              nativeBackgroundAndroid={
                {
                  "attribute": "selectableItemBackground",
                  "rippleRadius": undefined,
                  "type": "ThemeAttrAndroid",
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "borderRadius": 100,
                  "minWidth": 65,
                  "paddingHorizontal": 16,
                  "paddingVertical": 8,
                }
              }
              testID="backupQuiz/prosper"
            >
              <Text
                style={
                  {
                    "color": "#1AB775",
                    "textAlign": "center",
                  }
                }
              >
                prosper
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "borderColor": "#1AB775",
                "borderRadius": 100,
                "borderWidth": 1.5,
                "marginHorizontal": 4,
                "marginVertical": 4,
                "overflow": "hidden",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              focusable={true}
              nativeBackgroundAndroid={
                {
                  "attribute": "selectableItemBackground",
                  "rippleRadius": undefined,
                  "type": "ThemeAttrAndroid",
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "borderRadius": 100,
                  "minWidth": 65,
                  "paddingHorizontal": 16,
                  "paddingVertical": 8,
                }
              }
              testID="backupQuiz/winner"
            >
              <Text
                style={
                  {
                    "color": "#1AB775",
                    "textAlign": "center",
                  }
                }
              >
                winner
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "borderColor": "#1AB775",
                "borderRadius": 100,
                "borderWidth": 1.5,
                "marginHorizontal": 4,
                "marginVertical": 4,
                "overflow": "hidden",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              focusable={true}
              nativeBackgroundAndroid={
                {
                  "attribute": "selectableItemBackground",
                  "rippleRadius": undefined,
                  "type": "ThemeAttrAndroid",
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "borderRadius": 100,
                  "minWidth": 65,
                  "paddingHorizontal": 16,
                  "paddingVertical": 8,
                }
              }
              testID="backupQuiz/find"
            >
              <Text
                style={
                  {
                    "color": "#1AB775",
                    "textAlign": "center",
                  }
                }
              >
                find
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "borderColor": "#1AB775",
                "borderRadius": 100,
                "borderWidth": 1.5,
                "marginHorizontal": 4,
                "marginVertical": 4,
                "overflow": "hidden",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              focusable={true}
              nativeBackgroundAndroid={
                {
                  "attribute": "selectableItemBackground",
                  "rippleRadius": undefined,
                  "type": "ThemeAttrAndroid",
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "borderRadius": 100,
                  "minWidth": 65,
                  "paddingHorizontal": 16,
                  "paddingVertical": 8,
                }
              }
              testID="backupQuiz/donate"
            >
              <Text
                style={
                  {
                    "color": "#1AB775",
                    "textAlign": "center",
                  }
                }
              >
                donate
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "borderColor": "#1AB775",
                "borderRadius": 100,
                "borderWidth": 1.5,
                "marginHorizontal": 4,
                "marginVertical": 4,
                "overflow": "hidden",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              focusable={true}
              nativeBackgroundAndroid={
                {
                  "attribute": "selectableItemBackground",
                  "rippleRadius": undefined,
                  "type": "ThemeAttrAndroid",
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "borderRadius": 100,
                  "minWidth": 65,
                  "paddingHorizontal": 16,
                  "paddingVertical": 8,
                }
              }
              testID="backupQuiz/tape"
            >
              <Text
                style={
                  {
                    "color": "#1AB775",
                    "textAlign": "center",
                  }
                }
              >
                tape
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "borderColor": "#1AB775",
                "borderRadius": 100,
                "borderWidth": 1.5,
                "marginHorizontal": 4,
                "marginVertical": 4,
                "overflow": "hidden",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              focusable={true}
              nativeBackgroundAndroid={
                {
                  "attribute": "selectableItemBackground",
                  "rippleRadius": undefined,
                  "type": "ThemeAttrAndroid",
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "borderRadius": 100,
                  "minWidth": 65,
                  "paddingHorizontal": 16,
                  "paddingVertical": 8,
                }
              }
              testID="backupQuiz/history"
            >
              <Text
                style={
                  {
                    "color": "#1AB775",
                    "textAlign": "center",
                  }
                }
              >
                history
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
